File: app/src/main/java/com/techyourchance/mvc/screens/questionslist/QuestionsListAdapter.java
package com.techyourchance.mvc.screens.questionslist;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.TextView;

import com.techyourchance.mvc.R;
import com.techyourchance.mvc.questions.Question;

public class QuestionsListAdapter extends ArrayAdapter<Question> implements QuestionListItemViewMvc.Listener {

    private final OnQuestionClickListener mOnQuestionClickListener;

    public interface
    OnQuestionClickListener {
        void onQuestionClicked(Question question);
    }

    public QuestionsListAdapter(Context context,
                                OnQuestionClickListener onQuestionClickListener) {
        super(context, 0);
        mOnQuestionClickListener = onQuestionClickListener;
    }

    @NonNull
    @Override
    public View getView(int position, @Nullable View convertView, @NonNull ViewGroup parent) {
        if (convertView == null) {
            QuestionListItemViewMvc viewMvc = new QuestionListItemViewMvcImpl(
                    LayoutInflater.from(getContext()),parent
            );
            viewMvc.registerListener(this);
            convertView = viewMvc.getRootView();
            convertView.setTag(viewMvc);
        }

        final Question question = getItem(position);

        // bind the data to views
        QuestionListItemViewMvc viewMvc = (QuestionListItemViewMvc) convertView.getTag();
        viewMvc.bindQuestion(question);

        // set listener
        convertView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                onQuestionClicked(question);
            }
        });

        return convertView;
    }

    @Override
    public void onQuestionClicked(Question question) {
        mOnQuestionClickListener.onQuestionClicked(question);
    }

}

File: app/src/main/java/com/techyourchance/mvc/screens/questionslist/QuestionsListAdapter.kt
package com.techyourchance.mvc.screens.questionslist

import android.content.Context
import android.widget.ArrayAdapter
import com.techyourchance.mvc.questions.Question
import android.view.ViewGroup
import android.view.LayoutInflater
import android.view.View

class QuestionsListAdapter(
    context: Context?,
    private val mOnQuestionClickListener: OnQuestionClickListener
) : ArrayAdapter<Question?>(context!!, 0), QuestionListItemViewContract.Listener {
    interface OnQuestionClickListener {
        fun onQuestionClicked(question: Question?)
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        var convertView = convertView
        if (convertView == null) {
            val viewMvc: QuestionListItemViewContract = QuestionListItemView(
                LayoutInflater.from(context), parent
            )
            viewMvc.registerListener(this)
            convertView = viewMvc.rootView
            convertView!!.tag = viewMvc
        }
        val question = getItem(position)

        // bind the data to views
        val viewMvc = convertView.tag as QuestionListItemViewContract
        viewMvc.bindQuestion(question)

        // set listener
        convertView!!.setOnClickListener { onQuestionClicked(question) }
        return convertView
    }

    override fun onQuestionClicked(question: Question?) {
        mOnQuestionClickListener.onQuestionClicked(question)
    }
}